### PR TITLE
Cato 4629

### PR DESCRIPTION
**Pull request by wjrbetts**

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/ct/computations/CP3020.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      .map(new LocalDate(_, 2, 29))
+      .exists(l => !l.isBefore(apStart) && !l.isAfter(apEnd))
+
+    if (leapDayInAp) 366 else 365
+  }
+}
+
+
+object CP3020 {
+  val grassrootsStart = LocalDate.parse("2017-04-01")
```
Comment (wjrbetts) [inline src/main/scala/uk/gov/hmrc/ct/computations/CP3020.scala]: We could possibly move this but not a big deal.

File: src/main/scala/uk/gov/hmrc/ct/computations/CP999.scala [hunk 2]
```diff
@@ -16,17 +16,29 @@
...
 
-  override def calculate(fieldValueRetriever: ComputationsBoxRetriever): CP999 = {
-    calculateTotalDonations(cpq21 = fieldValueRetriever.cpQ21(),
-                            cp301 = fieldValueRetriever.cp301(),
-                            cp302 = fieldValueRetriever.cp302())
+  override def calculate(retriever: ComputationsBoxRetriever): CP999 = {
+    CP999(calculateNonGrassrootsDonations(retriever) + calculateGrassrootsDonations(retriever))
+  }
+
+  private def calculateNonGrassrootsDonations(retriever: ComputationsBoxRetriever) = {
```
Comment (wjrbetts) [inline src/main/scala/uk/gov/hmrc/ct/computations/CP999.scala]: Guess we could call this calculateCharitableDonations...

File: src/main/scala/uk/gov/hmrc/ct/computations/CPQ321.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.joda.time.LocalDate
+import uk.gov.hmrc.ct.box._
+import uk.gov.hmrc.ct.computations.Validators.DonationsValidation
+import uk.gov.hmrc.ct.computations.retriever.ComputationsBoxRetriever
+
+case class CPQ321(value: Option[Boolean]) extends CtBoxIdentifier(name = "Did your company make any donations to grassroots sports?")
+  with CtOptionalBoolean with Input with ValidatableBox[ComputationsBoxRetriever] with DonationsValidation {
+  override def validate(boxRetriever: ComputationsBoxRetriever): Set[CtValidation] = {
+    collectErrors(
+      failIf(!boxRetriever.cp2.value.isBefore(LocalDate.parse("2017-04-01")) && noValue) {
```
Comment (wjrbetts) [inline src/main/scala/uk/gov/hmrc/ct/computations/CPQ321.scala]: This should probably go off the same constant as (currently) defined in CP3020

File: src/main/scala/uk/gov/hmrc/ct/computations/CP3020.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    failIf(orZero > limit) {
+      Set(CtValidation(Some("CP3020"), "error.CP3020.apportionedLimit.exceeded", Some(Seq("£" + limit.toString))))
+    }
+  }
+
+  private def apportionedLimit(apStart: LocalDate, apEnd: LocalDate) = {
+    val daysAfter010417 = Days.daysBetween(grassrootsStart, apEnd).getDays + 1
+    val apDays = Days.daysBetween(apStart, apEnd).getDays + 1
+    val eligibleDays =  (daysAfter010417 min apDays) max 0
+    (eligibleDays * 2500) / daysInYear(apStart, apEnd)
```
Comment (Jrmcneil) [inline src/main/scala/uk/gov/hmrc/ct/computations/CP3020.scala]: we should use maxValue instead of hard coding 2500

**Pull request merged**